### DISCUSSION TITLE
How to keep the android app running in the background？

### DISCUSSION THREAD

**Discussion by ccinsz** [beeware/briefcase#1553] [Q&A]
I wrote a program about android app which it can keep running in the background like the music app.  But it run a while, it will suspend.  I'm eager to know how to implement it. 

This is similar code.

```
import threading
import time

import toga
from toga.style import Pack
from toga.style.pack import COLUMN


def test():
    for i in range(1000):
        print(i)
        time.sleep(10)


class demo1123(toga.App):
    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN))
        btn = toga.Button('Button', on_press=self.start)
        main_box.add(btn)

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

    def start(self, widget):
        threading.Thread(target=test).start()


def main():
    return demo1123()
```

**Comment by freakboy3742** [comment]
Mobile platforms don't behave the same as desktop platforms - you can't just start a thread and have it continue to run in the background while the app isn't in the foreground. I believe an Android activity *can* be configured to allow this but it's generally not advisable because it has a bad impact on battery usage. 

Android's documentation has [a detailed discussion of the ways you can deal with "background" activity](https://developer.android.com/guide/background/).

**Reply by ccinsz** [reply]
How to implement background service or foreground service in the android app by briefcase? Does it has any example code?

**Reply by freakboy3742** [reply]
It's not a feature that Briefcase currently supports, because it requires a bunch of Android-specific configuration. You'll need to reference the documentation I provided earlier, and integrate those changes into the Briefcase-generated template, using the Chaquopy layer to provide Python implementations of any required classes/interfaces.

**Reply by ccinsz** [reply]
I looked at some foreground service in Java code. It requires a service class. I don’t know how to convert it through Python in the Briefcase-generated template. Are there any related cases?

**Reply by freakboy3742** [reply]
No, I can't point you at examples of a feature that we don't support.

The documentation for Chaquopy can be found [here](https://chaquo.com/chaquopy/doc/current/python.html).

**Comment by maoist2009** [comment]
Have you find a solution? I want to make a local proxy and need to setup a foreground service too.

**Reply by freakboy3742** [reply]
As I noted in my previous reply, this isn't a feature that Briefcase supports at present. A background service is essentially a different kind of app. I'm guessing it's possible to implement it using Android APIs, but I can't point you at any examples of this, and I have no idea whether it would be plausible to add an API to Briefcase to support it.
